### PR TITLE
[SPARK-45887][SQL] Align codegen and non-codegen implementation of `Encode`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to change the implementation of interpretation mode, and make it consistent to codegen. Both implementation raise the same error with new error class `INVALID_PARAMETER_VALUE.CHARSET`.

### Why are the changes needed?
To make codegen and non-codegen of the `Encode` expression consistent. So, users will observe the same behaviour in both modes.

### Does this PR introduce _any_ user-facing change?
Yes, if user code depends on error from `encode()`.

### How was this patch tested?
By running the following test suites:
```
$ PYSPARK_PYTHON=python3 build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z string-functions.sql"
$ build/sbt "core/testOnly *SparkThrowableSuite"
$ build/sbt "test:testOnly *.StringFunctionsSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @srielau In the PR, I made the `Encode` implementation consistent. Could you review this PR, please.

File: common/utils/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -2024,6 +2024,11 @@
           "expects an integer value in [0, <upper>), but got <invalidValue>."
         ]
       },
+      "CHARSET" : {
+        "message" : [
+          "expects one of the charsets 'US-ASCII', 'ISO-8859-1', 'UTF-8', 'UTF-16BE', 'UTF-16LE', 'UTF-16', but got <charset>."
```
Comment (srielau) [inline common/utils/src/main/resources/error/error-classes.json:2029]: Should we give advice on the legacy configuration?

**Review by srielau — APPROVED** [review]

File: common/utils/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:2029]: I plan to restrict the supported charsets in the code, and add a config for the legacy behaviour. In the following PR, I will modify the message with an advice.

**Review by HyukjinKwon — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -2693,16 +2694,22 @@ case class Encode(value: Expression, charset: Expression)
...
+    } catch {
+      case _: UnsupportedEncodingException =>
+        throw QueryExecutionErrors.invalidCharsetError(prettyName, toCharset)
+    }
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     nullSafeCodeGen(ctx, ev, (string, charset) =>
       s"""
+        String toCharset = $charset.toString();
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2708]: It seems this is defined already.
```
[info] - SPARK-22543: split large if expressions into blocks due to JVM code size limit *** FAILED *** (59 milliseconds)
[info]   java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 145, Column 8: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 145, Column 8: Redefinition of local variable "toCharset"
```

**Review by dongjoon-hyun — COMMENTED** [review]
Could you make the CI happy?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2708]: Please use `val toCharset = ctx.freshName("toCharset")`

**Review by beliefer — APPROVED** [review]
LGTM except one comment.

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @dongjoon-hyun @srielau @cloud-fan @beliefer @HyukjinKwon for review.

**Pull request closed without merging**